ntfs_forensics/src/ntfsindxtable.cpp

```diff
@@ -98,59 +53,199 @@ void populateIndexRow(osquery::Row& r,
+    auto device_it = device_map.find(current_partition.device);
+    if (device_it == device_map.end()) {
+      device_map.insert(
+          {current_partition.device, {current_partition.part_address}});
 
-  if (devices.empty() || partitions.size() != 1) {
-    return {};
+    } else {
+      auto& partition_set = device_it->second;
+      partition_set.insert(current_partition.part_address);
+    }
+  }
+
+  // Get the SQL statement constraints
+  auto device_constraints =
+      request.constraints["device"].getAll(osquery::EQUALS);
+
+  if (device_constraints.empty()) {
+    for (const auto& p : device_map) {
+      const auto& device_name = p.first;
+      device_constraints.insert(device_name);
+    }
   }
 
-  std::stringstream part_stream;
-  int partition;
-  part_stream << *(partitions.begin());
-  part_stream >> partition;
-
-  for (const auto& dev : devices) {
-    trailofbits::Device* d = NULL;
-    trailofbits::Partition* p = NULL;
-    try {
-      d = new trailofbits::Device(dev);
-      p = new trailofbits::Partition(*d, partition);
-    } catch (std::runtime_error&) {
-      delete p;
-      delete d;
+  std::unordered_set<std::uint32_t> partition_constraints;
+  for (const auto& partition_str :
+       request.constraints["partition"].getAll(osquery::EQUALS)) {
+    char* null_term_ptr = nullptr;
+    auto partition_number =
+        std::strtoul(partition_str.c_str(), &null_term_ptr, 10);
+
+    if (*null_term_ptr != 0) {
+      VLOG(1) << "Invalid partition specified: " << partition_str
+              << ". Skipping...";
+
       continue;
     }
 
-    trailofbits::DirEntryList entries;
-    trailofbits::FileInfo fileInfo;
-    if (paths.size() == 1) {
-      p->collectINDX(std::string(*(paths.begin())), entries);
-      p->getFileInfo(*(paths.begin()), fileInfo);
-    } else if (inodes.size() == 1) {
-      std::stringstream inode_str;
-      uint64_t inode;
-      inode_str << *inodes.begin();
-      inode_str >> inode;
-      p->collectINDX(inode, entries);
-      p->getFileInfo(inode, fileInfo);
+    partition_constraints.insert(static_cast<std::uint32_t>(partition_number));
+  }
+
+  // Iterate through all devices
+  osquery::QueryData results;
+
+  for (const auto& device_name : device_constraints) {
+    // Make sure the specified device exists
+    auto device_map_it = device_map.find(device_name);
+    if (device_map_it == device_map.end()) {
+      VLOG(1) << "Device " << device_name << " was not found. Skipping...";
+      continue;
     }
 
-    for (auto& entry : entries) {
-      osquery::Row r;
-      populateIndexRow(r, entry, dev, partition, fileInfo.path);
-      results.push_back(r);
+    const auto& available_device_partitions = device_map_it->second;
+
+    // Iterate through all partitions
+    std::unordered_set<std::uint32_t> partition_list;
+    if (partition_constraints.empty()) {
+      partition_list = available_device_partitions;
+    }
```

Comment:
this clause ensures that a query that specifies a partition (e.g., "WHERE partition=2") will cause partition_list to always be empty and thus no partitions get checked

Reply:
Yeah I noticed, but while I was working on that file I also noticed that the other table needed the same thing so I halted that and started working on a more unified solution! I should have checked both files first!